src/atc/cache/CachedLoader.py

```diff
@@ -0,0 +1,215 @@
+        return None
+
+    def save(self, df: DataFrame) -> None:
+        in_cols = df.columns
+
+        if not set(self.params.key_cols).issubset(in_cols):
+            raise AssertionError(
+                "The key columns must be given in the input dataframe."
+            )
+
+        # ensure that no duplicates exist. The keys have to be a unique set.
+        df = df.dropDuplicates(self.params.key_cols)
+
+        cache = self._extract_cache()
+
+        result = self._discard_non_new_rows_against_cache(df, cache)
+
+        # write branch
+        df_written = self.write_operation(result.to_be_written)
+        if df_written:
+            # this line only executes if the line above does not raise an exception.
+            df_written_cache_update = self._prepare_written_cache_update(
+                df_written, in_cols
+            )
+            self._load_cache(df_written_cache_update)
+
+        # delete branch
+        df_deleted = self.delete_operation(result.to_be_deleted)
+        if df_deleted:
+            df_deleted_cache_update = self._prepare_deleted_cache_update(df_deleted)
+
+            self._load_cache(df_deleted_cache_update)
+            # this method is called a again separately in order to ensure that the
+            # written cache is saved even if the delete operation fails.
+
+        return
+
+    def _extract_cache(self) -> DataFrame:
+        # here we fix the version,
+        # so we don't overwrite the cache before we might want to use it.
+        version = (
+            Spark.get()
+            .sql(f"DESCRIBE HISTORY {self.params.cache_table_name} LIMIT 1")
+            .select("version")
+            .take(1)[0][0]
+        )
+        cache = Spark.get().sql(
+            f"SELECT * FROM {self.params.cache_table_name} VERSION AS OF {version}"
+            f" WHERE {self.params.deletedTime} IS NULL"
+            f" AND {self.params.rowHash} IS NOT NULL"
+            f" AND {self.params.loadedTime} IS NOT NULL"
+        )
+
+        return cache
+
+    def _prepare_written_cache_update(self, df: DataFrame, columns_to_hash: List[str]):
+        # re-hash the input row
+        return df.select(
+            *self.params.key_cols,
+            f.hash(*columns_to_hash).alias(self.params.rowHash),
+            f.current_timestamp().alias(self.params.loadedTime),
+            f.lit(None).cast("timestamp").alias(self.params.deletedTime),
+            *self.params.cache_id_cols,
+        )
+
+    def _prepare_deleted_cache_update(self, df: DataFrame):
+        return df.select(
+            *self.params.key_cols,
+            self.params.rowHash,
+            self.params.loadedTime,
+            f.current_timestamp().alias(self.params.deletedTime),
+            *self.params.cache_id_cols,
+        )
+
+    def _load_cache(self, cache_to_load: DataFrame) -> None:
+
+        view_name = (
+            self.params.cache_table_name.replace(".", "_")
+            + f"_Update{randint(0, 1000):03}"
+        )
```

Comment:
In atc.function there is a "get_unique_tempview_name()" you could consider to use.

Reply:
imlemented.